libcxx/test/support/test_macros.h

```diff
@@ -12,15 +12,15 @@
 
 // Attempt to get STL specific macros like _LIBCPP_VERSION using the most
 // minimal header possible. If we're testing libc++, we should use `<__config>`.
-// If <__config> isn't available, fall back to <ciso646>.
+// If <__config> isn't available, fall back to <version>.
 #ifdef __has_include
 # if __has_include("<__config>")
 #   include <__config>
 #   define TEST_IMP_INCLUDED_HEADER
 # endif
 #endif
 #ifndef TEST_IMP_INCLUDED_HEADER
-#include <ciso646>
+#  include <version>
```

Comment:
The libc++ test suite is intended to work with other standard libraries and all Standard versions. This change might break some Standard library in C++03 mode. Can you use something like. (please make sure this example is formatted properly)

```
#ifndef TEST_IMP_INCLUDED_HEADER
#  ifdef __has_include
#    if __has_include("<version>")
#     include <version>
#     define TEST_IMP_INCLUDED_HEADER
#   endif
# endif
#endif
#ifndef TEST_IMP_INCLUDED_HEADER
#  include <ciso646>
#endif
```

This should keep the test suite portable.

Reply:
TBH I don't think that makes a lot of sense. Yes, this makes the tests technically a tiny bit less portable, but every standard library should backport `<version>` anyways.

Reply:
I guess this comes down to the question of whether the test suite really can and should be fully independent of the implementation it's supposed to test. I certainly don't have any insight into where the libc++ test suite stands on that. If the test suite really only tests the library interface defined in the C++ specification, then I do see the point of making it work with any hypothetical implementation. That little added complexity would probably be justified then. I'll be happy to add that to the PR, I'll just wait for a clear decision on whether it should be added.

Reply:
Speaking of this: Is this whole
```cpp
#ifdef __has_include
# if __has_include("<__config>")
#   include <__config>
#   define TEST_IMP_INCLUDED_HEADER
# endif
#endif
```part at the top actually necessary? Shouldn't that stuff already be pulled in by `<version>` or `<ciso646>`?

If so, all of this could probably just become
```cpp
#if __has_include("<version>")
#  include <version>
#else
#  include <ciso646>
#endif
```
?

Reply:
The tests are supposed to work on any implementation - with the corresponding defect reports applied. Since there was guidance to support `<version>` in all standards modes from the committee, you could consider it to be a DR, which would mean we can use it for all modes.

Reply:
@philnik777 do you have a link for "Since there was guidance to support <version> in all standards modes from the committee", mainly since I'm curious.

Reply:
I wasn't able to find anything explicitly stating that (I might have missed it, since finding anything in the old meeting minutes is just awful), so I might be wrong that there was explicit guidance.

Reply:
FWIW I think the current patch is acceptable. The only way to not support `<version>` in C++03 mode is to explicitly add a `#error` in it based on the C++ version, and I'd be really surprised if any implementation did that.

Reply:
There are other ways to do it; the header is not required to be a file. However I'm not aware a Standard library not using files. I don't feel too strongly about this so we can leave this part of my change request. I still like the other change request addressed.

Reply:
I have now replaced this whole part
```cpp
// Attempt to get STL specific macros like _LIBCPP_VERSION using the most
// minimal header possible. If we're testing libc++, we should use `<__config>`.
// If <__config> isn't available, fall back to <ciso646>.
#ifdef __has_include
# if __has_include("<__config>")
#   include <__config>
#   define TEST_IMP_INCLUDED_HEADER
# endif
#endif
#ifndef TEST_IMP_INCLUDED_HEADER
#include <ciso646>
#endif
```
with just
```cpp
#ifdef __has_include
# if __has_include("<version>")
#   include <version>
# else
#   include <ciso646>
# endif
#else
#  include <ciso646>
#endif
```
It doesn't seem necessary to explicitly check and include `__config.h` since `<version.h>` already does that for libc++, so this seemed like a valid simplification?

Reply:
`<__config>` is only a libc++ thing, and we make no attempt to support another hypothetical library that would define that header. It is diligent from you to consider that case, but yeah that's not really a concern here.